### PR TITLE
Jocelyn july 13

### PR DESCRIPTION
**Pull request by jociet101**
# July 13 PR

# 👩🏻‍🏫 What does this PR do?

Friends feature
- [x] Split Friend requests vc into Incoming and outgoing requests
- [x] Allow user to cancel outgoing friend request
- [x] Improved UI of friend feature related views
- [x] Implemented friends list view controller and able to click on cell to view player profile

# Screenshots:

https://user-images.githubusercontent.com/73032138/178930862-3430ec52-0ca0-46f7-b0fd-ab60c9a2195e.mov

### CODE REVIEW

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m [hunk 1]
```diff
@@ -86,7 +93,6 @@ -(void)manageFriendButtonUI {
             self.isMyFriend = NO;
             [self setRequestPendingAsState];
         }
-        return;
     }
     
     [query whereKey:@"userObjectId" equalTo:self.user.objectId];
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:98]: IMO, keys as `userObjectId` are the ones that you want to have as const

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m [hunk 2]
```diff
@@ -86,7 +93,6 @@ -(void)manageFriendButtonUI {
             self.isMyFriend = NO;
             [self setRequestPendingAsState];
         }
-        return;
```
Comment (albertomota-codes) [inline Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:89]: why did you remove the return? 
It was a bug?

File: Playmate/Views/FriendCell.m [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
+    [super setSelected:selected animated:animated];
+
+    // Configure the view for the selected state
+}
+
+- (void)setThisUserId:(NSString *)thisUserId {
+    _thisUserId = thisUserId;
+    
+    NSLog(@"self.thisuserid %@", self.thisUserId);
```
Comment (albertomota-codes) [inline Playmate/Views/FriendCell.m:33]: Do we want to remove this?

File: Playmate/Views/OutgoingRequestCell.m [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    PFQuery *query = [PFUser query];
+    PFUser *user = [query getObjectWithId:userObjectId];
+    [user fetchIfNeeded];
+    
+    NSString *name = [Constants concatenateFirstName:user[@"firstName"][0] andLast:user[@"lastName"][0]];
+    self.titleLabel.text = [@"You requested to be friends with " stringByAppendingString:name];
+    
+    self.cancelButton.layer.cornerRadius = [Constants smallButtonCornerRadius];
+    
+    if (user[@"profileImage"] != nil) {
```
Comment (albertomota-codes) [inline Playmate/Views/OutgoingRequestCell.m:46]: We could also have:
```
const BOOL hasProfileImage = user[@"profileImage"] != nil;
UIImage* img = [UIImage imageNamed:hasProfileImage ?  user[@"profileImage"]: @"playmate_logo_transparent.png"];
```

File: Playmate/Views/OutgoingRequestCell.m [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        UIImage* img = [UIImage imageNamed:@"playmate_logo_transparent.png"];
+        [self.profileImageView setImage:[Constants resizeImage:img withDimension:83]];
+    }
+    self.profileImageView.layer.cornerRadius = self.profileImageView.frame.size.width/2.0f;
+}
+
+- (IBAction)cancelOutgoingRequest:(id)sender {
+    
+    PFUser *user = [PFUser currentUser];
+    [user fetchIfNeeded];
```
Comment (albertomota-codes) [inline Playmate/Views/OutgoingRequestCell.m:60]: why do we always need to call  fetchIfNeed? 
Should we just put it inside `[PFUser currentUser]`

File: Playmate/Views/OutgoingRequestCell.m [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    NSMutableArray *tempPendingList = (NSMutableArray *)connection[@"pendingList"];
+    [tempPendingList removeObject:self.userObjectId];
+    connection[@"pendingList"] = (NSArray *)tempPendingList;
+    
+    [connection saveInBackground];
+    
+    // delete the friend request object
+    PFQuery *query = [PFQuery queryWithClassName:@"FriendRequest"];
+    [query whereKey:@"requestFromId" equalTo:user.objectId];
+    [query whereKey:@"requestToId" equalTo:self.userObjectId];
```
Comment (albertomota-codes) [inline Playmate/Views/OutgoingRequestCell.m:74]: I would put the keys and  Classname in the Constant.h file.

File: Playmate/Utilities/Constants.h [hunk 7]
```diff
@@ -63,13 +63,22 @@ NS_ASSUME_NONNULL_BEGIN
 + (NSString *)emptyRequestsPlaceholderMsg;
 + (NSString *)emptyRequestsPlaceholderTitle;
 
++ (NSString *)emptyOutgoingRequestsPlaceholderTitle;
+
+// Information for friend list vc
++ (NSString *)emptyListPlaceholderMsg;
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/Utilities/Constants.h:69]: I was unaware that lazy constant properties are a thing in objective c/swift, I'll be sure to look into it!

File: Playmate/View Controllers/Session Details View/PlayerProfileViewController.m (hunk 2 above)
Comment (jociet101) [reply Playmate/View Controllers/Session Details View/PlayerProfileViewController.m:89]: Yep, the return caused a bug since there is a corner case that was not accounted for in this if statement, so I had to allow it to continue. The purpose was for detecting whether the player's profile being viewed is friends with the logged in users.

File: Playmate/Views/OutgoingRequestCell.m (hunk 5 above)
Comment (jociet101) [reply Playmate/Views/OutgoingRequestCell.m:60]: Often when fetchIfNeeded is not called on a Parse PFObject, there are run time errors saying the object has no value and its values cannot be accessed. Could you please clarify what you mean by putting it inside? Do you mean likke this: `PFUser *user = [[PFUser currentUser] fetchIfNeeded]` ?

File: Playmate/View Controllers/Profile Tab/FriendsListViewController.m [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    
+    if ([segue.identifier isEqualToString:@"toProfile"]) {
+        PlayerProfileViewController *vc = [segue destinationViewController];
+        PFQuery *query = [PFUser query];
+        NSIndexPath *indexPath = [self.tableView indexPathForCell:sender];
+        PFUser *theFriend = [query getObjectWithId:self.friendsList[indexPath.row]];
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/View Controllers/Profile Tab/FriendsListViewController.m:105]: Yep, we are sure because we set the number of rows in the table view to be equal to friendlist count

File: Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+    // fetch data asynchronously
+    [query findObjectsInBackgroundWithBlock:^(NSArray *requests, NSError *error) {
+        if (requests != nil) {
+            
+            self.friendRequestList = requests;
+            
+            [self.tableView reloadData];
+        } else {
+            NSLog(@"%@", error.localizedDescription);
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m:59]: I use an external pod library for displaying a small title and message for when the table view is empty, so an empty table view is fine. The only issues might be network connection problems, and I plan on adding an alert controller for that later.

File: Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+- (NSArray *)copyArray:(NSArray *)array {
+    NSMutableArray *copyTo = [[NSMutableArray alloc] init];
+    
+    for (FriendRequest *item in array) {
+        [copyTo addObject:item];
+    }
+    
+    return (NSArray *)copyTo;
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m:71]: Also realized that I don't even need this method so I cleaned it up

File: Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    return [[NSAttributedString alloc] initWithString:text attributes:attributes];
+}
+
+- (NSAttributedString *)descriptionForEmptyDataSet:(UIScrollView *)scrollView
+{
+    NSString *text = [Constants emptyRequestsPlaceholderMsg];
+    
+    NSMutableParagraphStyle *paragraph = [NSMutableParagraphStyle new];
+    paragraph.lineBreakMode = NSLineBreakByWordWrapping;
+    paragraph.alignment = NSTextAlignmentCenter;
```
[the remark this answers is not in the document]
Comment (jociet101) [reply Playmate/View Controllers/Profile Tab/IncomingRequestsViewController.m:118]: Yep just did!
